FLUID-6177: Adds a selection reader to the orator

jobara opened the pull request:
Addresses:
https://issues.fluidproject.org/browse/FLUID-6175
https://issues.fluidproject.org/browse/FLUID-6176
https://issues.fluidproject.org/browse/FLUID-6177

Includes changes from https://github.com/fluid-project/infusion/pull/898

jobara commented:
@amb26 could you please review this PR?

gtirloni commented:
@jobara could you please merge .buildkite/pipeline.yml from master? This should fix the build errors.

jobara commented:
@gtirloni thanks, I've updated with master and all the checks seem to be passing now.

jobara commented:
@amb26 could you please review this PR. It includes all of the changes to update the self voicing preference.

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+                        }
+                    },
+                    listeners: {
+                        "utteranceOnEnd.domReaderStop": [{
```
No need for this to be an array since any more than one member of a namespaced listener would be discarded

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+        }
+    });
+
+    fluid.orator.controller.toggleState = function (that, path, state) {
```
Docs for this function given its requirements seem a bit loose - might be helpful to mention it is designed to be invoked directly as a click handler. What is the reason for the odd polymorphism based on accepting the state either from the direct argument or from the model?

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+                args: ["{that}", "{arguments}.0"]
+            },
+            play: {
+                funcName: "fluid.orator.domReader.speak",
```
Awkward mismatch here between the name of the invoker and its implementing function.

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+     *
+     * @return {String} - The parsed text combined into a String.
+     */
+    fluid.orator.domReader.parsedToString = function (parsed) {
```
This should also be an invoker on domReader since it is the inverse of one of its operations, and invoked that way

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+     * to map the synthesized speech to the location of each word in the DOM; which is useful for highlighting words as
+     * they are being read.
+     *
+     * @typedef {Array} DomWordMappings
```
On the one hand I don't think this needs to be a typedef since it is just an array, and on the other hand I think it could usefully be documented what it is actually an array of - that is, you should use {DomWordMap[]} as the @param type below

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+     * Returns the index of the closest data point from the parseQueue based on the boundary provided.
+     *
+     * @param {Component} that - The component
+     * @param {Integer} boundary - The boundary value used to compare against the blockIndex of the parsed data points.
```
Looks like this parameter is optional

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+    fluid.orator.domReader.getClosestIndex = function (that, boundary) {
+        var parseQueue = that.parseQueue;
+
+        if (!parseQueue.length || !fluid.isValue(boundary) && boundary <= 0) {
```
It can't both be not a value and <= 0

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+            return undefined;
+        }
+
+        while (index >= 0) {
```
Thanks, I think the workflow of this function is much more comprehensible now. Also it will be hugely more efficient :)

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+            return undefined;
+        };
+
+        var maxIndex  = Math.max(parseQueue.length - 1, 0);
```
Stray space here

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+                speaking: false,
+                enabled: true
+            },
+            parseQueuelength: 0,
```
Capitalising this as parseQueueLength feels nicer

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+            }
+        },
+        modelListeners: {
+            "parseIndex": {
```
As you mentioned in our call, this might also be invalidated by a change in parseQueueLength that does not at the same time change parseIndex. You can use the "path" member to listen to both at once: https://docs.fluidproject.org/infusion/development/ChangeApplierAPI.html#the-path-entry-in-a-model-listener-declaration

amb26 commented on src/components/orator/js/Orator.js:
```diff
@@ -0,0 +1,963 @@
+    };
+
+    fluid.orator.domReader.removeExtraWhiteSpace = function (text) {
+
```
This function could do with some extra whitespace removed :)

amb26 commented on src/components/textToSpeech/js/TextToSpeech.js:
```diff
@@ -242,47 +288,149 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
+     *
+     * @param {Component} that - the component
+     * @param {Object} utteranceEventMap - a mapping from SpeechSynthesisUtterance events to component events.
+     * @param {Object} utteranceOpts - options to configure the SpeechSynthesis utterance with.
```
Looks like this could usefully be a @typedef since it appears in a few places

amb26 commented on src/framework/core/js/NewViewSupport.js:
```diff
@@ -0,0 +1,88 @@
+        invokers: {
+            renderMarkup: "fluid.identity({that}.options.markup.container)",
+            renderContainer: "fluid.containerRenderingView.renderContainer({that}, {that}.renderMarkup, {that}.addToParent)",
+            addToParent: {
```
Yup, I think this looks pretty good now and makes a fair amount of sense

jobara commented:
@amb26 thanks for the review. I believe I've addressed all of your comments. The PR is now ready for more review. (NOTE: as of writing this comment, the PR hadn't updated even though the changes have been pushed up to my branch.)
